### DISCUSSION TITLE
Patched Gnomoria fails to start on Linux

### DISCUSSION THREAD

**Discussion by TomFaulkner** [Nefaro/gnoll#7] [General]
I followed the instructions, though I had to move the game to get the installer to see it. (It won't show hidden, and Steam installs in `.local`.)

The game runs as expected when unpatched.

Output from the shell is below. The output remains the same if debug is set to true in the ini file.

I am attempting to run this on Arch / EndeavorOS.

Thanks!

```
033c:err:winediag:gnutls_process_attach failed to load libgnutls, no support for encryption
033c:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
033c:fixme:nls:GetFileMUIPath stub: 0x10, L"C:\\windows\\system32\\tzres.dll", (null), 006AEFB8, 031B45C0, 006AEFBC, 006AEFB0
033c:fixme:nls:GetFileMUIPath stub: 0x10, L"C:\\windows\\system32\\tzres.dll", (null), 006AEFB8, 031B45C0, 006AEFBC, 006AEFB0
033c:fixme:nls:GetFileMUIPath stub: 0x10, L"C:\\windows\\system32\\tzres.dll", (null), 006AEFB8, 031B45C0, 006AEFBC, 006AEFB0
(11/20/2023 5:53:19 PM) [Gnoll] == Gnoll Mod Loader 1.14.0 ==
(11/20/2023 5:53:20 PM) [Gnoll] Applying patches ...
(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.GnomanEmpi
re::PlayGame()
Unknown heap type: CLR

(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.GnomanEmpi
re::LoadGame(System.String fileName, System.Boolean fallenKingdom)
(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.GnomanEmpi
re::method_5(System.Boolean bool_1)
(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.GameSettin
gs::ApplyDisplayChanges()
(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.GUI.MainMe
nuWindow::.ctor(Game.GUI.Controls.Manager manager)
(11/20/2023 5:53:20 PM) [Gnoll] -- Patched System.Void Game.AIDirector
::Embark()
(11/20/2023 5:53:20 PM) [Gnoll] Applying patches ... DONE
* Assertion at /vagrant/mono/mono/metadata/metadata.c:1381, condition 
`idx < t->rows' not met

033c:err:mscoree:expect_no_runtimes Process exited with a Mono runtime loaded.
```

**Comment by TomFaulkner** [comment]
@aanjkyks, does the current version still work for you?

**Comment by Nefaro** [comment]
I must confess that I have 0 experience with mono on a linux machine.
What I can think of, maybe, is a version conflict with mono and the exe. I also don't have currently the means to test this out myself, however, for future reference, please tell me, what version of mono you are running.
Also, if you have time, you can try earlier versions, for example 1.12
1.13 added ingame mod management, which includes a json library, this is one of the suspects because the next line of the log should be something like (on windows):
`(21-Nov-23 22:24:05) [Gnoll] Loading mod states from C:\Users\<username>\Documents\My Games\Gnomoria\Gnoll\modlist.json`
So maybe the json library is breaking things.

**Reply by Nefaro** [reply]
Example of my debug log (for comparison)

> (21-Nov-23 22:24:04) [Gnoll] == Gnoll Mod Loader 1.14.0 ==
(21-Nov-23 22:24:04) [Gnoll] Applying patches ...
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.GnomanEmpire::PlayGame()
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.GnomanEmpire::LoadGame(System.String fileName, System.Boolean fallenKingdom)
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.GnomanEmpire::method_5(System.Boolean bool_1)
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.GameSettings::ApplyDisplayChanges()
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.GUI.MainMenuWindow::.ctor(Game.GUI.Controls.Manager manager)
(21-Nov-23 22:24:05) [Gnoll] -- Patched System.Void Game.AIDirector::Embark()
(21-Nov-23 22:24:05) [Gnoll] Applying patches ... DONE
(21-Nov-23 22:24:05) [Gnoll] Loading mod states from C:\Users\[username]\Documents\My Games\Gnomoria\Gnoll\modlist.json
(21-Nov-23 22:24:05) [Gnoll] Checking 'Gnoll Mods\enabled\BuildListCache.dll' for mods ...
(21-Nov-23 22:24:05) [Gnoll] ++ Instantiating mod: BuildListCache
(21-Nov-23 22:24:05) [Gnoll] Checking 'Gnoll Mods\enabled\Challenges.dll' for mods ...
(21-Nov-23 22:24:05) [Gnoll] ++ Instantiating mod: Challenges
(21-Nov-23 22:24:05) [Gnoll] Checking 'Gnoll Mods\enabled\ExpLuaIntegration.dll' for mods ...
(21-Nov-23 22:24:05) [Gnoll] ++ Instantiating mod: __ExpLuaIntegration
(21-Nov-23 22:24:05) [Gnoll] -- Mod has scripts
(21-Nov-23 22:24:05) [Gnoll] Checking 'Gnoll Mods\enabled\FasterEntityProcessing.dll' for mods ...
 ... ... ...

**Reply by TomFaulkner** [reply]
That's a great question. I don't have Mono installed on my system, but it's whatever Proton is providing. I have tried multiple versions of Proton. Perhaps a line could be added to the debug output to include that information?

Gnoll 1.12 works. So, I think you are probably correct in that.

**Reply by Nefaro** [reply]
Alright, that at least narrows it down ... I will try to debug and analyze it one day, but for now I suspect all nixes are stuck at 1.12.

**Reply by TomFaulkner** [reply]
I see here that latest Proton uses Mono 8.0.1:
https://github.com/ValveSoftware/Proton/releases/tag/proton-8.0-4

**Reply by TomFaulkner** [reply]
I appreciate the help!

**Reply by TomFaulkner** [reply]
I tried 1.13.1 today, it works as expected. I don't know if that changes anything about what the suspected issue is.

Happy Thanksgiving!

**Reply by Nefaro** [reply]
Alright, thank you for letting me know. This helps to narrow it down to only the latest version. At some point I'll take a closer look and/or add more debugging code.

**Comment by Nefaro** [comment]
After thinking about this for over a year (as can be seen), I finally had time to work on this issue. Spent about a week trying to figure out what's wrong and how to fix it, unfortunately it seems to be an issue with Mono and the compilation used by one of the libraries I am using, to be precise, the library (MoonSharp) I use for Lua integration.

After a week I was ready to declare "Linux + Wine" not supported (I did investigate modding the native engine, but, that's even more work and ... I have forgotten much). 
However, luckily I stumbled upon another solution/workaround. The library offers several binaries for different targets and one of the binaries worked for Windows as well as Linux + Wine.

As such, as of now, the issue has a resolution and v1.14.1 should work with Linux + wine as well.

On the other hand I now need to remember to test the setup on my Linux env as well.

**Reply by TomFaulkner** [reply]
Perfect timing. I just started playing again yesterday! I'll give the new release a try.

**Reply by TomFaulkner** [reply]
It starts and seems to be working!

Thank you for fixing this!

Since you're around, do you happen to know how difficult it would be to change the maximum resolution? The game caps it to 2048, and if set higher in the ini file it will revert to a safe default.

**Reply by Nefaro** [reply]
Good to know, it's not only my machine where it works.

Changing the hard-coded resolution is quite of a complex issue. At some point I tried to figure out how to do that and just offer the list of resolutions that the display supports... however, in the end it was too complex and had already hacked quite a bit of code and it was still not working. 
As it currently stands I don't see myself revisiting that particular issue anytime soon.

**Reply by TomFaulkner** [reply]
Understandable. Thanks again!
